drivers/sensor/st/lsm6dsv16x/lsm6dsv16x_rtio_stream.c

```diff
@@ -0,0 +1,344 @@
+	}  else {
+		lsm6dsv16x_pin_int2_route_set(ctx, &pin_int);
+	}
+}
+
+void lsm6dsv16x_submit_stream(const struct device *dev, struct rtio_iodev_sqe *iodev_sqe)
+{
+	struct lsm6dsv16x_data *lsm6dsv16x = dev->data;
+	const struct sensor_read_config *cfg = iodev_sqe->sqe.iodev->data;
+	uint8_t fifo_irq = 0;
+
+	gpio_pin_interrupt_configure_dt(lsm6dsv16x->drdy_gpio, GPIO_INT_DISABLE);
+
+	for (size_t i = 0; i < cfg->count; i++) {
+		if ((cfg->triggers[i].trigger == SENSOR_TRIG_FIFO_WATERMARK) ||
+		    (cfg->triggers[i].trigger == SENSOR_TRIG_FIFO_FULL)) {
+			fifo_irq = 1;
+		}
+	}
+
+	/* if any change in fifo irq */
+	if (fifo_irq != lsm6dsv16x->fifo_irq) {
+		lsm6dsv16x->fifo_irq = fifo_irq;
+
+		/* enable/disable the FIFO */
+		lsm6dsv16x_config_fifo(dev, fifo_irq);
+	}
+
+	lsm6dsv16x->streaming_sqe = iodev_sqe;
+
+	gpio_pin_interrupt_configure_dt(lsm6dsv16x->drdy_gpio, GPIO_INT_EDGE_TO_ACTIVE);
+}
+
+static void lsm6dsv16x_complete_op_cb(struct rtio *r, const struct rtio_sqe *sqe, void *arg)
+{
+	const struct device *dev = arg;
+	struct lsm6dsv16x_data *lsm6dsv16x = dev->data;
+
+	/*
+	 * Mark operation completed
+	 */
+	rtio_iodev_sqe_ok(sqe->userdata, 0);
+	lsm6dsv16x->streaming_sqe = NULL;
+	gpio_pin_interrupt_configure_dt(lsm6dsv16x->drdy_gpio, GPIO_INT_EDGE_TO_ACTIVE);
+}
+
+static void lsm6dsv16x_read_fifo_cb(struct rtio *r, const struct rtio_sqe *sqe, void *arg)
+{
+	const struct device *dev = arg;
+	struct lsm6dsv16x_data *lsm6dsv16x = dev->data;
+	struct gpio_dt_spec *irq_gpio = lsm6dsv16x->drdy_gpio;
+	struct rtio_iodev *iodev = lsm6dsv16x->iodev;
+	struct sensor_read_config *read_config;
+	uint8_t fifo_th = 0, fifo_full = 0;
+	uint16_t fifo_count;
+
+	/* proceed only if streaming */
+	if (lsm6dsv16x->streaming_sqe == NULL) {
+		return;
+	}
+
+	read_config = (struct sensor_read_config *)lsm6dsv16x->streaming_sqe->sqe.iodev->data;
+
+	if (read_config == NULL) {
+		return;
+	}
+
+	if (read_config->is_streaming == false) {
+		return;
+	}
+
+	struct sensor_stream_trigger *fifo_ths_cfg = NULL;
+	struct sensor_stream_trigger *fifo_full_cfg = NULL;
+
+	for (int i = 0; i < read_config->count; ++i) {
+		if (read_config->triggers[i].trigger == SENSOR_TRIG_FIFO_WATERMARK) {
+			fifo_ths_cfg = &read_config->triggers[i];
+			continue;
+		}
+
+		if (read_config->triggers[i].trigger == SENSOR_TRIG_FIFO_FULL) {
+			fifo_full_cfg = &read_config->triggers[i];
+			continue;
+		}
+	}
+
+	/* fill fifo h/w status */
+	fifo_th = (lsm6dsv16x->fifo_status[1] & 0x80) ? 1 : 0;
+	fifo_full = (lsm6dsv16x->fifo_status[1] & 0x20) ? 1 : 0;
+	fifo_count = (uint16_t)lsm6dsv16x->fifo_status[1] & 0x1U;
+	fifo_count = (fifo_count << 8) | lsm6dsv16x->fifo_status[0];
+	lsm6dsv16x->fifo_count = fifo_count;
+
+	bool has_fifo_ths_trig = fifo_ths_cfg != NULL && fifo_th == 1;
+	bool has_fifo_full_trig = fifo_full_cfg != NULL && fifo_full == 1;
+
+	/* check for spurious interrupts */
+	if (!has_fifo_ths_trig && !has_fifo_full_trig) {
+		lsm6dsv16x->streaming_sqe = NULL;
```

Comment:
Will this leave the submission without a completion here? If so this will cause issues.

Reply:
Thanks for catching this out... I'll return completion ok though (not error). Seems to me more proper.